File: src/main/java/ink/pd2/shell/Shell.java
package ink.pd2.shell;

import ink.pd2.shell.core.Listener;
import ink.pd2.shell.core.Logger;
import ink.pd2.shell.core.Resources;
import ink.pd2.shell.plugin.CommandExecutedListener;

import java.util.ArrayList;

public class Shell {
	//常量
	public final static int DEFAULT_PRIORITY = 2;

	//Shell信息
	private String title = "Pd2 Shell";
	private String user;
	public String getUser() {
		return user;
	}

	public String getCurrentUser() {
		return Main.getVariableValue("user");
	}

	//分优先级的指令执行事件
	private final ArrayList<ArrayList<CommandExecutedListener>> listeners
			= new ArrayList<>(5);

	public Shell() {
		listeners.add(new ArrayList<>(0));
		listeners.add(new ArrayList<>(0));
		listeners.add(new ArrayList<>(3));
		listeners.add(new ArrayList<>(0));
		listeners.add(new ArrayList<>(1));
		Logger.INS.writeLog("Shell", Resources.INS.getString("psh.log-init-shell"));
		//将事件按优先级排序
		Listener[] ls = Resources.INS.getListeners("psh", "command-executed");
		for (Listener li : ls) {
			CommandExecutedListener l = (CommandExecutedListener) li;
			listeners.get(l.getPriority()).add(l);
		}
		user = getCurrentUser();
	}

	public void run() {
		shell:
		while (true) {
			//指令处理和执行
			String c = Main.input.getCommand();
			for (ArrayList<CommandExecutedListener> ls : listeners)
				for (CommandExecutedListener l : ls) {
				if (!
					l.event(this, c)
				) break shell;
			}

		}
		//退出事件

	}

}

File: src/main/kotlin/ink/pd2/shell/Shell.kt
package ink.pd2.shell

import ink.pd2.shell.log.writeLog
import ink.pd2.shell.core.Resources
import ink.pd2.shell.io.Input
import ink.pd2.shell.io.Output
import ink.pd2.shell.plugin.CommandExecutedListener

class Shell(val input: Input, val output: Output) {
	//常量
	companion object {
		const val DEFAULT_PRIORITY: Int = 2
	}

	//Shell信息
	var title: String = "Pd2 Shell"
	private var user: String
		get() = field
	fun getCurrentUser(): String {
		return getVariableValue("user")!!
	}

	//分优先级的指令执行事件
	private val listeners = arrayOf<ArrayList<CommandExecutedListener>>(
		ArrayList(1), ArrayList(1), ArrayList(10),
		ArrayList(1), ArrayList(2))

	init {
		writeLog("Shell", Resources.getString("psh.log-init-shell"))
		//将事件按优先级排序
		val ls = Resources.getListeners("psh", "command-executed")
		for (li in ls) {
			val l = li as CommandExecutedListener
			listeners[l.priority].add(l)
		}
		user = getCurrentUser()
	}

	fun run() {
		shell@
		while (true) {
			//指令处理和执行
			val c = input.command
			for (ls in listeners) for (l in ls) {
				if (!
					l.event(this, c)
				) break@shell
			}

		}
		//退出事件

	}

}
